augur/filter/include_exclude_rules.py

```diff
@@ -188,24 +189,34 @@ def filter_by_query(metadata: pd.DataFrame, query: str) -> FilterFunctionReturn:
     # Create a copy to prevent modification of the original DataFrame.
     metadata_copy = metadata.copy()
 
-    # Try extracting column names from the query.
-    variables = extract_variables(query)
-    if variables is not None:
-        columns = variables.intersection(metadata_copy.columns)
+    # Convert data types before applying the query.
+    if column_types:
+        # If types are explicitly provided, use those as source of truth.
+        for column, dtype in column_types.items():
+            if dtype == 'int' or dtype == 'float':
+                metadata_copy[column] = pd.to_numeric(metadata_copy[column], errors='ignore')
```

Comment:
I'd expect `int` or `float` to mean actually `int` or `float` and not mean "general numeric conversion". What do you think about allowing `numeric` to mean the `pd.to_numeric(…)` call (i.e. "int or float, whatever it is") and `int` or `float` to mean strict conversion (or maybe truncation) to those specific types? I think that would mean calling `.astype('int')` and `.astype('float')` (or maybe their nullable equivalents, `'Int64'` and `'Float64'`?), possibly with an `errors=…` kwarg too.

Relatedly, do we want to use `errors='ignore'` when the user specifies a specific type for the column? It's friendly in some cases but hostile in others. (Which points to making it specifiable…)

Reply:
That sounds better. Will update.

Reply:
Updated in d84c47cc495be54501fdbbdaa62da6539be86834:

https://github.com/nextstrain/augur/blob/679519ee5092df34f7658b4e677bb8c5b87f69d2/augur/filter/include_exclude_rules.py#L213-L228

Reply:
I chose `errors=raise` since it's user error. Added a test to show what it looks like:

https://github.com/nextstrain/augur/blob/1a2dae9928e72b943dd213552a80bd9392f35b8c/tests/functional/filter/cram/filter-query-columns.t#L47-L56